near-sdk/src/store/lookup_map/entry.rs

```diff
@@ -0,0 +1,379 @@
+use crate::utils::CacheEntry;
+
+/// A view into a single entry in the map, which can be vacant or occupied.
+pub enum Entry<'a, K: 'a, V: 'a> {
+    Occupied(OccupiedEntry<'a, K, V>),
+    Vacant(VacantEntry<'a, K, V>),
+}
+
+impl<'a, K, V> Entry<'a, K, V>
+where
+    K: Ord,
+{
+    /// Returns a reference to this entry's key.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use near_sdk::store::LookupMap;
+    ///
+    /// let mut map: LookupMap<String, u32> = LookupMap::new(b"m");
+    /// assert_eq!(map.entry("poneyland".to_string()).key(), &"poneyland");
+    /// ```
+    pub fn key(&self) -> &K {
+        match self {
+            Entry::Occupied(entry) => entry.key(),
+            Entry::Vacant(entry) => entry.key(),
+        }
+    }
+
+    /// Ensures a value is in the entry by inserting the default if empty, and returns
+    /// a mutable reference to the value in the entry.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use near_sdk::store::LookupMap;
+    ///
+    /// let mut map: LookupMap<String, u32> = LookupMap::new(b"m");
+    ///
+    /// map.entry("poneyland".to_string()).or_insert(3);
+    /// assert_eq!(map["poneyland"], 3);
+    ///
+    /// *map.entry("poneyland".to_string()).or_insert(10) *= 2;
+    /// assert_eq!(map["poneyland"], 6);
+    /// ```
+    pub fn or_insert(self, default: V) -> &'a mut V {
+        match self {
+            Self::Occupied(entry) => entry.into_mut(),
+            Self::Vacant(entry) => entry.insert(default),
+        }
+    }
+
+    /// Ensures a value is in the entry by inserting the result of the default function if empty,
+    /// and returns a mutable reference to the value in the entry.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use near_sdk::store::LookupMap;
+    ///
+    /// let mut map: LookupMap<String, String> = LookupMap::new(b"m");
+    /// let s = "hoho".to_string();
+    ///
+    /// map.entry("poneyland".to_string()).or_insert_with(|| s);
+    ///
+    /// assert_eq!(map["poneyland"], "hoho".to_string());
+    /// ```
+    pub fn or_insert_with<F: FnOnce() -> V>(self, default: F) -> &'a mut V {
+        match self {
```

Comment:
this could delegate to `or_insert_with_key`

Reply:
Does this get compiled away? I just noticed that `std::collections::HashMap` [does not do this](https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#2791) and decided to follow their pattern.

Reply:
Yes, according to my mental model, the compiler should trivially compile that away. This pattern is used by `Option` and once cell: https://github.com/rust-lang/rust/blob/cdeba02ff71416e014f7130f75166890688be986/library/core/src/option.rs#L1228

It's interesting that maps don't do that. I suppose that's historical -- this is the form the API was added before 1.0, and the form it was copy-pasted into hash brown: https://github.com/rust-lang/rust/pull/22930. 

No strong opinion here though -- if you'll feel more confident with the current impl, that's totally fine!

Reply:
No, that's fine, I'll switch. I'm just noting the reason I had was that I assumed because it was the pattern in std::.::HashMap that there was a benefit